### PR TITLE
[SPARK-31981][SQL] Keep TimestampType when taking an average of a Timestamp

### PR DESCRIPTION
**Pull request by Fokko**
Currently, when you take an average of a Timestamp, you'll end up with a Double, representing the seconds since epoch. This is because of old Hive behavior.

I strongly believe that it is better to return a Timestamp, to make the behavior more congruent with the max function, for example:
```
scala> df.agg(max("ts"), avg("ts")).show()
+-------------------+----------+
|            max(ts)|   avg(ts)|
+-------------------+----------+
|2018-03-04 06:11:40|7.126971E8|
+-------------------+----------+
```

This might also improve performance because we get rid of the implicit cast.

Behaviour in Postgres:
```
root@8c4241b617ec:/# psql postgres postgres
psql (12.3 (Debian 12.3-1.pgdg100+1))
Type "help" for help.

postgres=# CREATE TABLE timestamp_demo (ts TIMESTAMP);
CREATE TABLE
postgres=# INSERT INTO timestamp_demo VALUES('2019-01-01 18:22:11');
INSERT 0 1
postgres=# INSERT INTO timestamp_demo VALUES('2018-01-01 18:22:11');
INSERT 0 1
postgres=# INSERT INTO timestamp_demo VALUES('2017-01-01 18:22:11');
INSERT 0 1
postgres=# SELECT AVG(ts) FROM timestamp_demo;
ERROR:  function avg(timestamp without time zone) does not exist
LINE 1: SELECT AVG(ts) FROM timestamp_demo;
```

Behaviour in MySQL:

```
root@bab43a5731e8:/# mysql
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 9
Server version: 8.0.20 MySQL Community Server - GPL

Copyright (c) 2000, 2020, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> CREATE TABLE timestamp_demo (ts TIMESTAMP);
Query OK, 0 rows affected (0.05 sec)

mysql> INSERT INTO timestamp_demo VALUES('2019-01-01 18:22:11');
Query OK, 1 row affected (0.01 sec)

mysql> INSERT INTO timestamp_demo VALUES('2018-01-01 18:22:11');
Query OK, 1 row affected (0.01 sec)

mysql> INSERT INTO timestamp_demo VALUES('2017-01-01 18:22:11');
Query OK, 1 row affected (0.01 sec)

mysql> SELECT AVG(ts) FROM timestamp_demo;
+---------------------+
| AVG(ts)             |
+---------------------+
| 20180101182211.0000 |
+---------------------+
1 row in set (0.00 sec)
```

Which is a YYYYMMDDHHMMSS format in double.

### What changes were proposed in this pull request?

Return a Timestamp when taking the average of a Timestamp column.

### Why are the changes needed?
I believe that it is awkward to implicitly change the type when doing an average.

### Does this PR introduce _any_ user-facing change?

Yes, it will return a Timestamp instead of a Double when taking an average of a Timestamp column.

### How was this patch tested?

Using unit tests and existing tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestData.scala [hunk 1]
```diff
@@ -73,6 +74,17 @@ private[sql] trait SQLTestData { self =>
     df
   }
 
+  protected lazy val timestampData: DataFrame = {
+    val df = spark.sparkContext.parallelize(
+      TestDataTimestamp(new Timestamp(1420140300000L)) :: // 2015-01-01 20:25:00
+        TestDataTimestamp(new Timestamp(1320140300000L)) :: // 2011-11-01 10:38:20
+        TestDataTimestamp(new Timestamp(1520140300000L)) :: // 2018-03-04 06:11:40
+        TestDataTimestamp(new Timestamp(-1409632500000L)) :: // 1925-05-01 19:44:32
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestData.scala:82]: Do you set the fractional part of seconds to zeros intentionally?

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Hi, @Fokko . The last failure means that we need to regenerate the output of `ThriftServerQueryTestSuite.sql`. To proceed more, please update it.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by Fokko** [comment]
Thanks for the restart @dongjoon-hyun. Another error now; let me dive into it and get back to y'all.

File: sql/core/src/test/resources/sql-tests/results/udf/udf-window.sql.out [hunk 2]
```diff
@@ -154,17 +154,17 @@ SELECT val_timestamp, udf(cate), avg(val_timestamp) OVER(PARTITION BY udf(cate)
...
-2017-08-17 13:00:00	b	1.503E9
-2020-12-30 16:00:00	b	1.6093728E9
+2017-07-31 17:00:00	NULL	2017-07-31 17:00:00
+2017-07-31 17:00:00	a	2017-08-02 11:04:26.666666
+2017-07-31 17:00:00	a	2017-08-02 11:04:26.666666
+2017-08-05 23:13:20	a	2017-08-05 23:13:20
+2020-12-30 16:00:00	a	2020-12-30 16:00:00
+2017-07-31 17:00:00	b	2017-08-09 03:00:00
+2017-08-17 13:00:00	b	2017-08-17 13:00:00
+2020-12-30 16:00:00	b	2020-12-30 16:00:00
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/results/udf/udf-window.sql.out:167]: Hm .. so we allow timestamp types whereas other DBMSes disallow?

Comment (Fokko) [reply sql/core/src/test/resources/sql-tests/results/udf/udf-window.sql.out:167]: Well, there isn't a real consensus around other DBMS'es. Keeping it a Timestamp seems like something that you would expect. MySQL's behavior is much more awkward in my opinion. Spark needs to pave the path on this one :)

File: sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestData.scala (hunk 1 above)
Comment (Fokko) [reply sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestData.scala:82]: No, this wasn't intentional. I can add some fractions if you like.

**Comment by Fokko** [comment]
Found the issue. `(sum / count).cast(resultType)` needs to be `(sum / count.cast(DecimalType.LongDecimal)).cast(resultType)` similar to the line above. The for the division operation, both the dividend and divisor requires to be a Fractional type.

File: sql/core/src/test/resources/sql-tests/results/udf/udf-window.sql.out (hunk 2 above)
Comment (maropu) [reply sql/core/src/test/resources/sql-tests/results/udf/udf-window.sql.out:167]: It seems Oracle cannot accept timestamp for average, too. Any other system supporting this behaivour?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 3]
```diff
@@ -624,7 +624,6 @@ object TypeCoercion {
 
       // Hive lets you do aggregation of timestamps... for some reason
       case Sum(e @ TimestampType()) => Sum(Cast(e, DoubleType))
-      case Average(e @ TimestampType()) => Average(Cast(e, DoubleType))
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:627]: When chaning the existing bheivour, we need to update the migration guide and might need to add a legacy config to keep the current behaivour.

**Pull request closed without merging**